src/octopath.cpp

```diff
@@ -4,92 +4,139 @@
-    static double epsilon = 0.0001, dist;
-    static int delete_index;
-    static Eigen::Vector3d p1, p2, p;
-    while (true) {
-        delete_index = -1;
-        for (int i = 1; i < compressed_points-1; i++) {
-            p1 << (*compressed_path)[i-1][0],
-                  (*compressed_path)[i-1][1],
-                  (*compressed_path)[i-1][2];
-            p2 << (*compressed_path)[i+1][0],
-                  (*compressed_path)[i+1][1],
-                  (*compressed_path)[i+1][2];
-            p  << (*compressed_path)[i][0],
-                  (*compressed_path)[i][1],
-                  (*compressed_path)[i][2];
-            algebra_3d::Line3d line(p1, p2);
-            line.DistancePoint2Line(p, &dist);
-            if (dist < epsilon) {
-                delete_index = i;
-                break;
-            }
-        }
-        if (delete_index > 0) {
-            compressed_path->erase(compressed_path->begin() + delete_index);
-            compressed_points = compressed_points - 1;
-        } else {
-            break;
-        }
+  // first delete colinear points
+  this->DeleteColinearWaypoints(path, compressed_path);
+  uint compressed_points = compressed_path->size();
+
+  // Compress the remaining points
+  static uint min_points = 2;
+  static bool add_final_waypoint = true;
+  static bool do_not_add_final_waypoint = false;
+  double dist, max_dist;
+  int delete_index, col_check;
+  Eigen::Vector3d p1, p2, p;
+  while (compressed_points > min_points) {
+    // first find the point that deviates the least in the whole set
+    max_dist = -1;
+    delete_index = -1;
+    for (uint i = 1; i < compressed_points-1; i++) {
+      p1 << (*compressed_path)[i-1][0],
+            (*compressed_path)[i-1][1],
+            (*compressed_path)[i-1][2];
+      p2 << (*compressed_path)[i+1][0],
+            (*compressed_path)[i+1][1],
+            (*compressed_path)[i+1][2];
+      p  << (*compressed_path)[i][0],
+            (*compressed_path)[i][1],
+            (*compressed_path)[i][2];
+      algebra_3d::Line3d line(p1, p2);
+      line.DistancePoint2Line(p, &dist);
+
+      // Check for collision
+      if (free_space_only) {
+        col_check = CheckCollision(p1, p2);
+      } else {
+        col_check = CheckOccupancy(p1, p2);
+      }
+
+      if (col_check == 1) {
+        continue;
+      }
+
+      // If doesn't collide, check if new nodes get closer to walls
+      // We don't allow to remove waypoints that decrease the distance to walls
+      std::vector<octomap::point3d> p1_to_p2_nodes, p1_to_p_to_p2_nodes;
+      this->GetNodesBetweenWaypoints(p1, p2, add_final_waypoint, &p1_to_p2_nodes);
+      this->GetNodesBetweenWaypoints(p1, p, do_not_add_final_waypoint, &p1_to_p_to_p2_nodes);
+      this->GetNodesBetweenWaypoints(p, p2, add_final_waypoint, &p1_to_p_to_p2_nodes);
+      if (this->AverageObstaclePathCost(p1_to_p2_nodes) > this->AverageObstaclePathCost(p1_to_p_to_p2_nodes)) {
+        continue;
+      }
```

Comment:
this is the only new portion within `PathPruning`. This forbids deleting intermediate waypoints that make the compressed set of waypoints closer to walls